domain/externalcontroller/state/state.go

```diff
@@ -168,51 +166,74 @@ func (st *State) ImportExternalControllers(ctx context.Context, infos []crossmod
 
 func (st *State) updateExternalControllerTx(
 	ctx context.Context,
-	tx *sql.Tx,
+	tx *sqlair.TX,
 	ci crossmodel.ControllerInfo,
 ) error {
 	cID := ci.ControllerTag.Id()
-	q := `
+	updateControllerStmt, err := sqlair.Prepare(`
 INSERT INTO external_controller (uuid, alias, ca_cert)
-VALUES (?, ?, ?)
-  ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert`
+VALUES ($ExternalController.uuid, $ExternalController.alias, $ExternalController.ca_cert)
+  ON CONFLICT(uuid) DO UPDATE SET alias=excluded.alias, ca_cert=excluded.ca_cert
+`, ExternalController{})
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	if _, err := tx.ExecContext(ctx, q, cID, ci.Alias, ci.CACert); err != nil {
+	externalController := ExternalController{
+		ID:     cID,
+		Alias:  sql.NullString{String: ci.Alias, Valid: true},
+		CACert: ci.CACert,
+	}
+	if err := tx.Query(ctx, updateControllerStmt, externalController).Run(); err != nil {
 		return errors.Trace(err)
 	}
 
-	addrsBinds, addrsAnyVals := database.SliceToPlaceholder(ci.Addrs)
-	q = fmt.Sprintf(`
+	deleteStmt, err := sqlair.Prepare(`
 DELETE FROM external_controller_address
-WHERE  controller_uuid = ?
-AND    address NOT IN (%s)`[1:], addrsBinds)
-
-	args := append([]any{cID}, addrsAnyVals...)
-	if _, err := tx.ExecContext(ctx, q, args...); err != nil {
+WHERE  controller_uuid = $ExternalController.uuid
+AND    address NOT IN ($ControllerUUIDs[:])
+`, ExternalController{}, ControllerUUIDs{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cIDs := ControllerUUIDs(ci.Addrs)
+	if len(cIDs) == 0 {
+		cIDs = ControllerUUIDs{}
+	}
+	if err := tx.Query(ctx, deleteStmt, externalController, cIDs).Run(); err != nil {
 		return errors.Trace(err)
 	}
 
-	for _, addr := range ci.Addrs {
-		q := `
+	insertAddressStmt, err := sqlair.Prepare(`
 INSERT INTO external_controller_address (uuid, controller_uuid, address)
-VALUES (?, ?, ?)
-  ON CONFLICT(controller_uuid, address) DO NOTHING`
-
-		if _, err := tx.ExecContext(ctx, q, utils.MustNewUUID().String(), cID, addr); err != nil {
+VALUES ($ExternalControllerAddress.uuid, $ExternalControllerAddress.controller_uuid,
+        $ExternalControllerAddress.address)
+  ON CONFLICT(controller_uuid, address) DO NOTHING
+`, ExternalControllerAddress{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, addr := range ci.Addrs {
+		externContAddr := ExternalControllerAddress{ID: utils.MustNewUUID().String(), ControllerUUID: cID, Addr: addr}
```

Comment:
whoa, missed this one in the review prior to your changes. Can you lift this `utils.MustNewUUID()` out of this call and call `utils.NewUUID()`. The former panics and we don't want that to happen in production.